feat(pack): `vim.pack.get()` gets VCS info

echasnovski opened the pull request:
This PR addresses two issue.

---

`spec.version` on `master` is explicitly resolved (`nil` values are replaced with default branch) when triggering events and inside `get()`. This was done to allow users and wrapper plugins to know which actual branch is used. However, this comes at consistency (`PackChangedPre` for install can't have it inferred, for example) and performance cost. More importantly, it also hides the information about whether the user explicitly set `version` or not, which might matter in some use cases (like detecting if `version` has actually changed).

This PR stops doing that and tries to preserve `spec` in state as close to what user (eventually) set it to. I.e. it converts string `src` to table field and infers `name` (because it is required for further operations).

---

To allow users and wrapper plugins to compute information about installed plugins, this PR updates `get()` to allow to return data only for a handful of plugins and opt in for computing extra info about them. The API is `get(names, opts)` with `opts.info` being `true` by default.

Two examples of usages:

- Get current state: `get({ 'plug-name' })[1].rev`      
- Get default branch: `get({ 'plug-name' })[1].branches[1]`

echasnovski committed:
fix(pack)!: do not resolve `spec.version` in event-data and `get()`

Problem: Force resolve `spec.version` overrides the information about
  whether a user supplied `version` or not. Knowing it might be useful
  in some use cases (like comparing to previously set `spec` to detect
  if it has changed).

Solution: Do not resolve `spec.version`. This also improves speed when
  triggering events and calling `get()`.

echasnovski committed:
refactor(pack): improve Git helpers

Details:
- Place default branch first when listing all branches.
- Use correct terminology in `get_hash` helper.
- Do not return `{ '' }` if there are no tags.

echasnovski committed:
feat(pack): make `get()` more capable

Problem: There is no way to get more information about installed
  plugins, like current revision or default branch (necessary if
  resolving default `spec.version` manually).
  As computing Git data migth take some time, also allow `get()` to
  limit output to only necessary set of plugins.

Solution: introduce arguments to `get(names, opts)`, which follows other
  `vim.pack` functions. Plugin extra info is returned by default and
  should be opt-out via `opts.info = false`.
  Examples:
  - Get current revision: `get({ 'plug-name' })[1].rev`
  - Get default branch: `get({ 'plug_name' })[1].branches[1]`

echasnovski committed:
refactor(pack): use `get(names)` to compute plug list from `names`

It also makes `update()` and `del()` to act on plugins in the same
order their names are supplied. This seems like a less surprising
behavior.

justinmk commented on runtime/doc/pack.txt:
```diff
@@ -355,16 +356,27 @@ del({names})                                                  *vim.pack.del()*
           to current session.
+        • {details}? (`{ branches: string[], state: string, tags: string[] }`)
+          • `branches` lists available Git branches (first is default).
+          • `state` is a current Git state.
```
example? is it a status string, or a dict? (showing an example or type in the this doc will answer this question for users)

echasnovski replied:
> s it a status string, or a dict? (showing an example or type in the this doc will answer this question for users)

It is shown higher that it is a string.



> if `state` is the commit, and never something like HEAD or a non-sha name, probably should just call it `rev` or `commit`. Git calls that a commit-id, but that's probably too pedantic. I lean towards `rev` because that's a bit more generalized but not too vague.

Currently it is indeed a (short) commit hash. Naming it with not VCS related name is a bit more forward compatible. Like after resolving #34765 a plugin can be local and can have its own notion of "state".

If taking only Git-related stuff, then It can be `rev`, I guess.

justinmk replied:
> Like after resolving #34765 a plugin can be local and can have its own notion of "state".

`rev` might work for that? What would be examples of local "state" values?

echasnovski replied:
> `rev` might work for that?

Saying that "local plugin's revision is xxx" just doesn't feel right. But maybe I am reading too much into this.

> What would be examples of local "state" values?

Don't know, but at least leaving the door open shouldn't be a bad thing. Plus somewhat related: documentation at start of `*vim.pack*` and confirmation buffer already use "State" / "State before"  / "State after". Which was chosen specifically to be more future compatible. It can also be changed, of course, but these need at least to be consistent.

justinmk commented on runtime/doc/pack.txt:
```diff
@@ -355,16 +356,27 @@ del({names})                                                  *vim.pack.del()*
+      • {names}  (`string[]?`) List of plugin names. Default: all plugins
+                 managed by |vim.pack|.
+      • {opts}   (`table?`) A table with the following fields:
+                 • {details} (`boolean`) Whether to include plugin details.
```
Can be 3 flags `branch`, `tags`, etc. That avoids the urge to put that info into a dict. Or name it `scm` or `vcs`, and it controls VCS-related fields (but they are not nested).

echasnovski replied:
On one hand, this can allow more precise control over performance (only enable flags which you want to see).

On the other hand, won't it become more confusing if the number of provided "extra" details grow? For example, there will probably also be `type` field if/when there are "local plugins". Plus the reasonable set of useful flags might be different for different plugin types.

Weighing both options, I'd probably still go with a single `details` / `info` flag that populates all extra available data about the plugin. For Git plugins they are computed in parallel, so shouldn't be too much of a problem. But I'd go the way you suggest eventually, of course.

echasnovski replied:
> True. In #30086 we are considering (the opposite concept) `fast`. And git has the concept of `sparse` checkouts.

I mean, it can be `fast` also. This is mostly the main reason of the flag's existence anyways: allow a way to control whether there needs to be time/resources spent to gather more information. Useful for both users and `vim.pack` itself (the `fast = true` or `info = false` case).

If `fast` is in consideration, then I'd go with it.

justinmk replied:
idk if `fast` is a good pattern to proliferate, because (1) performance is variable, (2) it has lower semantic value

let's say `info` for now

echasnovski commented on runtime/doc/pack.txt:
```diff
@@ -355,16 +356,27 @@ del({names})                                                  *vim.pack.del()*
-get()                                                         *vim.pack.get()*
-    Get data about all plugins managed by |vim.pack|
+get({names}, {opts})                                          *vim.pack.get()*
+    Get data about plugins managed by |vim.pack|
```
It is not (currently) filtered by `opts`. Should it mention that it can be filtered by `name`?

justinmk replied:
This doc is pretty sparse (not a bad thing), so I think mentioning "optionally filtered by..." is worthwhile.

echasnovski commented on runtime/doc/pack.txt:
```diff
@@ -355,16 +356,27 @@ del({names})                                                  *vim.pack.del()*
         • {path} (`string`) Plugin's path on disk.
         • {active} (`boolean`) Whether plugin was added via |vim.pack.add()|
           to current session.
+        • {details}? (`{ branches: string[], state: string, tags: string[] }`)
```
> ... is the nesting actually necessary? The spec lives in `spec` so that's already isolated[1]. `branches` and `tags` fit in top-level as much as `path`, no?

No, not really. Again, it mostly follows the `nvim_buf_get_extmarks()` approach, which returns all "extra" details in a single dictionary.

justinmk replied:
well `nvim_buf_get_extmarks` is wrong. it has like 73 fields so not a good example of anything, except at least it resisted the urge to create 73 functions.

sounds like we are aligned that these fields can be top-level.

echasnovski committed:
fixup(pack): address review comments

Details:
- Rename `opts.details` to `opts.info`.
- Set extra info directly in plugin info without nesting.

Also changed default `opts.info` to `true` since this simplifies logic
for the common user, while still leaving the door open for a faster
`get()` if needed.

echasnovski committed:
fixup(pack): for refactor to use `get(names)` directly

echasnovski commented:
Addressed review comments:
- Rename `opts.details` to `opts.info`.
- Set extra info directly in plugin data table without nesting.

Also changed default `opts.info` to `true` since this simplifies logic for the common user, while still leaving the door open for a faster `get()` if needed.

---

One thing to also consider might be to make `opts.info` private (maybe for now) with `opts` documented as "reserved for future use" (as it is still needed for `update()` and `del()`). The justification for this might be:
- With lockfile in place the `rev` field *might* be set from the cached lockfile data and won't need to be guarded behind `opts.info`. Not sure if this is a good idea for general design reasons: should `get()` always return *real* extra info (like now) or allowing it to go through cached data here is fine? I plan to [simplify part of getting non-active plugins](https://github.com/neovim/neovim/blob/13b04207f167b957bc8967c32d816349be024882/runtime/lua/vim/pack.lua#L1014-L1022) by using cached lockfile data, so that might already answer the question in favor of the latter.
- Simplifies documentation and API.

This all makes sense only if `opts._info=true` is default, though.

justinmk commented:
> With lockfile in place the `rev` field _might_ be set from the cached lockfile data and won't need to be guarded behind `opts.info`
> ... should `get()` always return _real_ extra info (like now) or allowing it to go through cached data here is fine?

I like the idea of using the cache by default and having a way to opt-in to "no cache", controlled by a flag (`nocache`, `resolve`). That is clearer than the current `info` flag that we discussed.

That doesn't need to block this, can revisit.

echasnovski commented:
> I like the idea of using the cache by default and having a way to opt-in to "no cache", controlled by a flag (`nocache`, `resolve`). That is clearer than the current `info` flag that we discussed.

It won't really be possible without actually storing `branches` and `tags` data on disk. They feel unnecessary in the "public" lockfile, while I *think* it is reasonable to avoid creating a separate "private" lockfile/cache for this kind of data.

justinmk commented:
> avoid creating a separate "private" lockfile/cache for this kind of data.

Oh, definitely a good idea to avoid that. Querying is better than caching, except where the performance cost is just too high.

justinmk commented:
> > With lockfile in place the `rev` field _might_ be set from the cached lockfile data and won't need to be guarded behind `opts.info`
> > ... should `get()` always return _real_ extra info (like now) or allowing it to go through cached data here is fine?
> 
> I like the idea of using the cache by default and having a way to opt-in to "no cache", controlled by a flag (`nocache`, `resolve`). That is clearer than the current `info` flag that we discussed.

For reference, the [docker client](https://github.com/neovim/neovim/blob/d017f3c9a0b745e0c57feb8c92dcc852948f7301/src/gen/gen_help_html.lua#L385) calls this `sparse`, that seems like a good name.
